cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,621 @@
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// Default to adding a precise match if possible. Otherwise
+	// let seccomp figure out the architecture specifics.
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	// note that all architecture strings are in the dpkg
+	// architecture notation
+	var compatArch string
+
+	// common case: kernel and userspace have the same arch. We
+	// add a compat architecture for some architectures that
+	// support it, e.g. on amd64 kernel and userland, we add
+	// compat i386 syscalls.
+	if arch.UbuntuArchitecture() == arch.UbuntuKernelArchitecture() {
+		switch arch.UbuntuArchitecture() {
+		case "amd64":
+			compatArch = "i386"
+		case "arm64":
+			compatArch = "armhf"
+		case "ppc64":
+			compatArch = "powerpc"
+		}
+	} else {
+		// less common case, kernel and userspace
+		// different. This can happen when running e.g. a
+		// amd64 kernel with a i386 userland. However in this
+		// case snapd would also only request i386 snaps. So
+		// the use-case is even stranger (from a distro
+		// perspective), a i386 snap would have to ship some
+		// 64bit code that would have to detect at runtime if
+		// it can be used or not. But we are told this is not
+		// rare with certain classes of embedded devices.
+		compatArch = arch.UbuntuKernelArchitecture()
+	}
+
+	// actually add the compat architecture (if we need one)
+	switch compatArch {
+	case "amd64":
+		return secFilter.AddArch(seccomp.ArchAMD64)
+	case "arm64":
+		return secFilter.AddArch(seccomp.ArchARM64)
+	case "ppc64el":
+		return secFilter.AddArch(seccomp.ArchPPC64LE)
+	case "i386":
+		return secFilter.AddArch(seccomp.ArchX86)
+	case "armhf":
+		return secFilter.AddArch(seccomp.ArchARM)
+	case "powerpc":
+		return secFilter.AddArch(seccomp.ArchPPC)
+	}
```

Comment:
In the C implementation we set initialized compat to `0` and then set it to `SCMP_ARCH_...` in the if/else up above and then check for if compat > 0. It would be easier to read if in the if/else here you did something similar and use `var compatArch ScmpArch` and set it to seccomp.ArchWHATEVER in the if/else. I think this would make it easier to read but it also means that this section of code is easier to maintain (just update the if/else as opposed to both the if/else and the switch). Indeed, you forgot to add `ppc64` to the switch even though you reference it in the if/else.